### PR TITLE
Checker on GAE

### PR DESCRIPTION
**Pull request by k2wanko**
- It works with GAE.
- Add SlackWebhookNotifier

please review.

### CODE REVIEW

File: backend/server.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	hc := urlfetch.Client(ctx)
+	var wg sync.WaitGroup
+	for _, s := range nschecker.Sources {
+		wg.Add(1)
+		go func(s nschecker.Source) {
+			defer wg.Done()
+			err := check(ctx, hc, s)
```
Comment (haya14busa) [inline backend/server.go]: [nit] `if err := check(ctx, hc, s); err != nil {`

File: backend/server.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func newNotifier(ctx context.Context, hc *http.Client) (nschecker.Notifier, error) {
+	tok := os.Getenv("SLACK_API_TOKEN")
+	if tok != "" {
+		channel := os.Getenv("SLACK_CHANNEL")
+		if channel == "" {
+			return nil, errors.New("Please set enviroment variable SLACK_CHANNEL")
+		}
+
```
Comment (haya14busa) [inline backend/server.go]: [nit] needless blank line?

File: cmd/nintendo-switch-checker/main.go [hunk 3]
```diff
@@ -221,6 +79,19 @@ func main() {
...
 		n = nschecker.NewLineNotifier(http.DefaultClient, tok)
+	case "slack-webhook":
+		url := os.Getenv("SLACK_WEBHOOK_URL")
+		if url == "" {
+			log.Println("Please set environment variable SLACK_WEBHOOK_URL")
+			return
+		}
+
+		if *channel == "" {
+			log.Println("Please set -slack-channel flag")
```
Comment (haya14busa) [inline cmd/nintendo-switch-checker/main.go]: s/-slack-channel/-channel/ or please change cli flag

File: nschecker/checker.go [hunk 4]
```diff
@@ -40,8 +40,27 @@ func (s State) String() string {
...
+	hc *http.Client
+}
+
+func HTTPClient(hc *http.Client) CheckOption {
+	return func(o *checkOptions) {
+		o.hc = hc
+	}
+}
+
+func Check(s Source, opts ...CheckOption) (State, error) {
```
Comment (haya14busa) [inline nschecker/checker.go]: hmm... i'm not sure it's good interface...

I think just passing HTTP.Client is fine. Use DefaultClient if it's nil.

File: nschecker/notify.go [hunk 5]
```diff
@@ -20,18 +18,19 @@ type Notifier interface {
 
 // SlackNotifier struct is construct an slack message.
```
Comment (haya14busa) [inline nschecker/notify.go]: I didn't notice it before but "is construct" is not valid English, i guess.

"SlackNotifier handles notification to slack" or something is better.

File: nschecker/notify.go [hunk 6]
```diff
@@ -9,8 +9,6 @@ import (
 	"net/url"
 
 	"strings"
-
-	"github.com/nlopes/slack"
 )
 
 // Notifier interface is construct an response.
```
Comment (haya14busa) [inline nschecker/notify.go]: it does not only construct an response.

It creates request and send them  to notification service.

```
// Notifier is interface of  notification
```
or something is fine

File: nschecker/notify.go [hunk 7]
```diff
@@ -58,8 +57,24 @@ func (n *SlackNotifier) Notify(state State, s Source) error {
 		channel = "<!channel|channel> "
 	}
 	msg := fmt.Sprintf("%s%v: %v (%v)", channel, state, s.URL, s.Name)
-	params := slack.PostMessageParameters{EscapeText: false}
-	_, _, err := n.Cli.PostMessage(n.channel, msg, params)
+	v := url.Values{}
```
Comment (haya14busa) [inline nschecker/notify.go:60]: why did you change SlackNotifier?

Please add comment or describe the reason in commit message.

File: nschecker/notify.go [hunk 8]
```diff
@@ -117,3 +132,64 @@ func (n *LineNotifier) Notify(state State, s Source) error {
 
 	return nil
 }
+
+// SlackWebhookNotifier struct is construct an slack incoming webhook message.
```
Comment (haya14busa) [inline nschecker/notify.go]: [ditto]

**Review by haya14busa — COMMENTED** [review]
Thanks :bird: :+1:

**Comment by haya14busa** [comment]
I heard that some source doesn't work in GAE.
Can you describe it about this? 
If it doesn't work, I think we need to filter the source.

File: nschecker/notify.go (hunk 7 above)
Comment (k2wanko) [reply nschecker/notify.go:60]: `github.com/nlopes/slack` was not build in go1.6.
GAE/Go is only go1.6.

**Comment by k2wanko** [comment]
https://github.com/haya14busa/nintendo-switch-checker/pull/3#issuecomment-309956749

@haya14busa 
I do not know the reason 😭 
will split source.

log:

```
Check failed: Yodobashi - Nintendo Switch Joy-Con(L)/(R)グレー [Nintendo Switch本体]: Get http://www.yodobashi.com/product/100000001003431565/: API error 12 (urlfetch: CONNECTION_ERROR)
Check failed: Yodobashi - Nintendo Switch Joy-Con(L)ネオンブルー/(R)ネオンレッド [Nintendo Switch本体]: Get http://www.yodobashi.com/product/100000001003431566/: API error 12 (urlfetch: CONNECTION_ERROR)
Check failed: Joshin - Nintendo Switch 本体【Joy-Con(L) ネオンブルー/(R) ネオンレッド】: Get http://joshinweb.jp/game/40519/4902370535716.html: API error 5 (urlfetch: DEADLINE_EXCEEDED)
Check failed: Joshin - Nintendo Switch 本体【Joy-Con(L)/(R) グレー】: Get http://joshinweb.jp/game/40519/4902370535709.html: API error 5 (urlfetch: DEADLINE_EXCEEDED)
```

**Comment by k2wanko** [comment]
@haya14busa please review.

File: backend/server.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+func init() {
+	http.HandleFunc("/", handler)
+}
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	hc := urlfetch.Client(ctx)
+	var wg sync.WaitGroup
+	for _, s := range nschecker.Sources {
```
Comment (haya14busa) [inline backend/server.go]: Please filter sources here instead of splitting the sources.

It's exceptional cases for GAE and should be handled here.

Leaving some comments also will help.

**Comment by haya14busa** [comment]
Can you change  backend/ to appengine/, app/, server/ or something if you come up with a good term?

**Comment by k2wanko** [comment]
@haya14busa changed

**Comment by haya14busa** [comment]
Please address this comment. https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123229832

**Comment by k2wanko** [comment]
@haya14busa oops 
please check.

**Review by haya14busa — APPROVED** [review]
:bird: :+1:
